{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "355c209d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import hail as hl\n",
    "import pandas as pd\n",
    "# Parameters\n",
    "log_path = '../logs/hail'\n",
    "file_in = '../data/merged.vcf.gz'\n",
    "file_out = '../data/merged.mt'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9a63a21",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "    <style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"c7a6142c-7431-49e7-ae25-314fa19a42a5\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "'use strict';\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    function drop(id) {\n      const view = Bokeh.index.get_by_id(id)\n      if (view != null) {\n        view.model.document.clear()\n        Bokeh.index.delete(view)\n      }\n    }\n\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n\n    // Clean up Bokeh references\n    if (id != null) {\n      drop(id)\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim()\n            drop(id)\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded(error = null) {\n    const el = document.getElementById(\"c7a6142c-7431-49e7-ae25-314fa19a42a5\");\n    if (el != null) {\n      const html = (() => {\n        if (typeof root.Bokeh === \"undefined\") {\n          if (error == null) {\n            return \"BokehJS is loading ...\";\n          } else {\n            return \"BokehJS failed to load.\";\n          }\n        } else {\n          const prefix = `BokehJS ${root.Bokeh.version}`;\n          if (error == null) {\n            return `${prefix} successfully loaded.`;\n          } else {\n            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n          }\n        }\n      })();\n      el.innerHTML = html;\n\n      if (error != null) {\n        const wrapper = document.createElement(\"div\");\n        wrapper.style.overflow = \"auto\";\n        wrapper.style.height = \"5em\";\n        wrapper.style.resize = \"vertical\";\n        const content = document.createElement(\"div\");\n        content.style.fontFamily = \"monospace\";\n        content.style.whiteSpace = \"pre-wrap\";\n        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n        content.textContent = error.stack ?? error.toString();\n        wrapper.append(content);\n        el.append(wrapper);\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(() => display_loaded(error), 100);\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n      try {\n            for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n\n      } catch (error) {display_loaded(error);throw error;\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"c7a6142c-7431-49e7-ae25-314fa19a42a5\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/10/31 22:21:24 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "Running on Apache Spark version 3.5.7\n",
      "SparkUI available at http://pimenta:4040\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.136-c32f88309ab0\n",
      "LOGGING: writing to ../logs/hail/\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-10-31 22:21:30.525 Hail: INFO: scanning VCF for sortedness...\n",
      "2025-10-31 22:21:32.466 Hail: INFO: Coerced sorted VCF - no additional import work to do\n",
      "2025-10-31 22:21:34.076 Hail: INFO: wrote matrix table with 2000 rows and 0 columns in 1 partition to ../data/merged.mt\n",
      "2025-10-31 22:31:02.810 Hail: INFO: wrote matrix table with 2000 rows and 0 columns in 1 partition to ../data/merged.mt\n",
      "2025-10-31 22:44:40.151 Hail: INFO: merging 2 files totalling 458,3K...\n",
      "2025-10-31 22:44:40.160 Hail: INFO: while writing:\n",
      "    ../data/variant_stats.tsv\n",
      "  merge time: 7,899ms\n"
     ]
    }
   ],
   "source": [
    "hl.init(app_name = 'gnomAD', log = log_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "48cbf36c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import and write matrix table\n",
    "mt = hl.import_vcf(file_in, force_bgz=True, reference_genome='GRCh38')\n",
    "mt.write(file_out, overwrite=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "36f5470c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt = hl.read_matrix_table(file_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "91c9d427",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matrix table schema:\n",
      "----------------------------------------\n",
      "Global fields:\n",
      "    None\n",
      "----------------------------------------\n",
      "Column fields:\n",
      "    's': str\n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'locus': locus<GRCh38>\n",
      "    'alleles': array<str>\n",
      "    'rsid': str\n",
      "    'qual': float64\n",
      "    'filters': set<str>\n",
      "    'info': struct {\n",
      "        AC: array<int32>, \n",
      "        AN: int32, \n",
      "        AF: array<float64>, \n",
      "        grpmax: array<str>, \n",
      "        fafmax_faf95_max: array<float64>, \n",
      "        fafmax_faf95_max_gen_anc: array<str>, \n",
      "        AC_XX: array<int32>, \n",
      "        AF_XX: array<float64>, \n",
      "        AN_XX: int32, \n",
      "        nhomalt_XX: array<int32>, \n",
      "        AC_XY: array<int32>, \n",
      "        AF_XY: array<float64>, \n",
      "        AN_XY: int32, \n",
      "        nhomalt_XY: array<int32>, \n",
      "        nhomalt: array<int32>, \n",
      "        AC_afr_XX: array<int32>, \n",
      "        AF_afr_XX: array<float64>, \n",
      "        AN_afr_XX: int32, \n",
      "        nhomalt_afr_XX: array<int32>, \n",
      "        AC_afr_XY: array<int32>, \n",
      "        AF_afr_XY: array<float64>, \n",
      "        AN_afr_XY: int32, \n",
      "        nhomalt_afr_XY: array<int32>, \n",
      "        AC_afr: array<int32>, \n",
      "        AF_afr: array<float64>, \n",
      "        AN_afr: int32, \n",
      "        nhomalt_afr: array<int32>, \n",
      "        AC_amr_XX: array<int32>, \n",
      "        AF_amr_XX: array<float64>, \n",
      "        AN_amr_XX: int32, \n",
      "        nhomalt_amr_XX: array<int32>, \n",
      "        AC_amr_XY: array<int32>, \n",
      "        AF_amr_XY: array<float64>, \n",
      "        AN_amr_XY: int32, \n",
      "        nhomalt_amr_XY: array<int32>, \n",
      "        AC_amr: array<int32>, \n",
      "        AF_amr: array<float64>, \n",
      "        AN_amr: int32, \n",
      "        nhomalt_amr: array<int32>, \n",
      "        AC_asj_XX: array<int32>, \n",
      "        AF_asj_XX: array<float64>, \n",
      "        AN_asj_XX: int32, \n",
      "        nhomalt_asj_XX: array<int32>, \n",
      "        AC_asj_XY: array<int32>, \n",
      "        AF_asj_XY: array<float64>, \n",
      "        AN_asj_XY: int32, \n",
      "        nhomalt_asj_XY: array<int32>, \n",
      "        AC_asj: array<int32>, \n",
      "        AF_asj: array<float64>, \n",
      "        AN_asj: int32, \n",
      "        nhomalt_asj: array<int32>, \n",
      "        AC_eas_XX: array<int32>, \n",
      "        AF_eas_XX: array<float64>, \n",
      "        AN_eas_XX: int32, \n",
      "        nhomalt_eas_XX: array<int32>, \n",
      "        AC_eas_XY: array<int32>, \n",
      "        AF_eas_XY: array<float64>, \n",
      "        AN_eas_XY: int32, \n",
      "        nhomalt_eas_XY: array<int32>, \n",
      "        AC_eas: array<int32>, \n",
      "        AF_eas: array<float64>, \n",
      "        AN_eas: int32, \n",
      "        nhomalt_eas: array<int32>, \n",
      "        AC_fin_XX: array<int32>, \n",
      "        AF_fin_XX: array<float64>, \n",
      "        AN_fin_XX: int32, \n",
      "        nhomalt_fin_XX: array<int32>, \n",
      "        AC_fin_XY: array<int32>, \n",
      "        AF_fin_XY: array<float64>, \n",
      "        AN_fin_XY: int32, \n",
      "        nhomalt_fin_XY: array<int32>, \n",
      "        AC_fin: array<int32>, \n",
      "        AF_fin: array<float64>, \n",
      "        AN_fin: int32, \n",
      "        nhomalt_fin: array<int32>, \n",
      "        AC_mid_XX: array<int32>, \n",
      "        AF_mid_XX: array<float64>, \n",
      "        AN_mid_XX: int32, \n",
      "        nhomalt_mid_XX: array<int32>, \n",
      "        AC_mid_XY: array<int32>, \n",
      "        AF_mid_XY: array<float64>, \n",
      "        AN_mid_XY: int32, \n",
      "        nhomalt_mid_XY: array<int32>, \n",
      "        AC_mid: array<int32>, \n",
      "        AF_mid: array<float64>, \n",
      "        AN_mid: int32, \n",
      "        nhomalt_mid: array<int32>, \n",
      "        AC_nfe_XX: array<int32>, \n",
      "        AF_nfe_XX: array<float64>, \n",
      "        AN_nfe_XX: int32, \n",
      "        nhomalt_nfe_XX: array<int32>, \n",
      "        AC_nfe_XY: array<int32>, \n",
      "        AF_nfe_XY: array<float64>, \n",
      "        AN_nfe_XY: int32, \n",
      "        nhomalt_nfe_XY: array<int32>, \n",
      "        AC_nfe: array<int32>, \n",
      "        AF_nfe: array<float64>, \n",
      "        AN_nfe: int32, \n",
      "        nhomalt_nfe: array<int32>, \n",
      "        AC_non_ukb_XX: array<int32>, \n",
      "        AF_non_ukb_XX: array<float64>, \n",
      "        AN_non_ukb_XX: int32, \n",
      "        nhomalt_non_ukb_XX: array<int32>, \n",
      "        AC_non_ukb_XY: array<int32>, \n",
      "        AF_non_ukb_XY: array<float64>, \n",
      "        AN_non_ukb_XY: int32, \n",
      "        nhomalt_non_ukb_XY: array<int32>, \n",
      "        AC_non_ukb: array<int32>, \n",
      "        AF_non_ukb: array<float64>, \n",
      "        AN_non_ukb: int32, \n",
      "        nhomalt_non_ukb: array<int32>, \n",
      "        AC_non_ukb_afr_XX: array<int32>, \n",
      "        AF_non_ukb_afr_XX: array<float64>, \n",
      "        AN_non_ukb_afr_XX: int32, \n",
      "        nhomalt_non_ukb_afr_XX: array<int32>, \n",
      "        AC_non_ukb_afr_XY: array<int32>, \n",
      "        AF_non_ukb_afr_XY: array<float64>, \n",
      "        AN_non_ukb_afr_XY: int32, \n",
      "        nhomalt_non_ukb_afr_XY: array<int32>, \n",
      "        AC_non_ukb_afr: array<int32>, \n",
      "        AF_non_ukb_afr: array<float64>, \n",
      "        AN_non_ukb_afr: int32, \n",
      "        nhomalt_non_ukb_afr: array<int32>, \n",
      "        AC_non_ukb_amr_XX: array<int32>, \n",
      "        AF_non_ukb_amr_XX: array<float64>, \n",
      "        AN_non_ukb_amr_XX: int32, \n",
      "        nhomalt_non_ukb_amr_XX: array<int32>, \n",
      "        AC_non_ukb_amr_XY: array<int32>, \n",
      "        AF_non_ukb_amr_XY: array<float64>, \n",
      "        AN_non_ukb_amr_XY: int32, \n",
      "        nhomalt_non_ukb_amr_XY: array<int32>, \n",
      "        AC_non_ukb_amr: array<int32>, \n",
      "        AF_non_ukb_amr: array<float64>, \n",
      "        AN_non_ukb_amr: int32, \n",
      "        nhomalt_non_ukb_amr: array<int32>, \n",
      "        AC_non_ukb_asj_XX: array<int32>, \n",
      "        AF_non_ukb_asj_XX: array<float64>, \n",
      "        AN_non_ukb_asj_XX: int32, \n",
      "        nhomalt_non_ukb_asj_XX: array<int32>, \n",
      "        AC_non_ukb_asj_XY: array<int32>, \n",
      "        AF_non_ukb_asj_XY: array<float64>, \n",
      "        AN_non_ukb_asj_XY: int32, \n",
      "        nhomalt_non_ukb_asj_XY: array<int32>, \n",
      "        AC_non_ukb_asj: array<int32>, \n",
      "        AF_non_ukb_asj: array<float64>, \n",
      "        AN_non_ukb_asj: int32, \n",
      "        nhomalt_non_ukb_asj: array<int32>, \n",
      "        AC_non_ukb_eas_XX: array<int32>, \n",
      "        AF_non_ukb_eas_XX: array<float64>, \n",
      "        AN_non_ukb_eas_XX: int32, \n",
      "        nhomalt_non_ukb_eas_XX: array<int32>, \n",
      "        AC_non_ukb_eas_XY: array<int32>, \n",
      "        AF_non_ukb_eas_XY: array<float64>, \n",
      "        AN_non_ukb_eas_XY: int32, \n",
      "        nhomalt_non_ukb_eas_XY: array<int32>, \n",
      "        AC_non_ukb_eas: array<int32>, \n",
      "        AF_non_ukb_eas: array<float64>, \n",
      "        AN_non_ukb_eas: int32, \n",
      "        nhomalt_non_ukb_eas: array<int32>, \n",
      "        AC_non_ukb_fin_XX: array<int32>, \n",
      "        AF_non_ukb_fin_XX: array<float64>, \n",
      "        AN_non_ukb_fin_XX: int32, \n",
      "        nhomalt_non_ukb_fin_XX: array<int32>, \n",
      "        AC_non_ukb_fin_XY: array<int32>, \n",
      "        AF_non_ukb_fin_XY: array<float64>, \n",
      "        AN_non_ukb_fin_XY: int32, \n",
      "        nhomalt_non_ukb_fin_XY: array<int32>, \n",
      "        AC_non_ukb_fin: array<int32>, \n",
      "        AF_non_ukb_fin: array<float64>, \n",
      "        AN_non_ukb_fin: int32, \n",
      "        nhomalt_non_ukb_fin: array<int32>, \n",
      "        AC_non_ukb_mid_XX: array<int32>, \n",
      "        AF_non_ukb_mid_XX: array<float64>, \n",
      "        AN_non_ukb_mid_XX: int32, \n",
      "        nhomalt_non_ukb_mid_XX: array<int32>, \n",
      "        AC_non_ukb_mid_XY: array<int32>, \n",
      "        AF_non_ukb_mid_XY: array<float64>, \n",
      "        AN_non_ukb_mid_XY: int32, \n",
      "        nhomalt_non_ukb_mid_XY: array<int32>, \n",
      "        AC_non_ukb_mid: array<int32>, \n",
      "        AF_non_ukb_mid: array<float64>, \n",
      "        AN_non_ukb_mid: int32, \n",
      "        nhomalt_non_ukb_mid: array<int32>, \n",
      "        AC_non_ukb_nfe_XX: array<int32>, \n",
      "        AF_non_ukb_nfe_XX: array<float64>, \n",
      "        AN_non_ukb_nfe_XX: int32, \n",
      "        nhomalt_non_ukb_nfe_XX: array<int32>, \n",
      "        AC_non_ukb_nfe_XY: array<int32>, \n",
      "        AF_non_ukb_nfe_XY: array<float64>, \n",
      "        AN_non_ukb_nfe_XY: int32, \n",
      "        nhomalt_non_ukb_nfe_XY: array<int32>, \n",
      "        AC_non_ukb_nfe: array<int32>, \n",
      "        AF_non_ukb_nfe: array<float64>, \n",
      "        AN_non_ukb_nfe: int32, \n",
      "        nhomalt_non_ukb_nfe: array<int32>, \n",
      "        AC_non_ukb_raw: array<int32>, \n",
      "        AF_non_ukb_raw: array<float64>, \n",
      "        AN_non_ukb_raw: int32, \n",
      "        nhomalt_non_ukb_raw: array<int32>, \n",
      "        AC_non_ukb_remaining_XX: array<int32>, \n",
      "        AF_non_ukb_remaining_XX: array<float64>, \n",
      "        AN_non_ukb_remaining_XX: int32, \n",
      "        nhomalt_non_ukb_remaining_XX: array<int32>, \n",
      "        AC_non_ukb_remaining_XY: array<int32>, \n",
      "        AF_non_ukb_remaining_XY: array<float64>, \n",
      "        AN_non_ukb_remaining_XY: int32, \n",
      "        nhomalt_non_ukb_remaining_XY: array<int32>, \n",
      "        AC_non_ukb_remaining: array<int32>, \n",
      "        AF_non_ukb_remaining: array<float64>, \n",
      "        AN_non_ukb_remaining: int32, \n",
      "        nhomalt_non_ukb_remaining: array<int32>, \n",
      "        AC_non_ukb_sas_XX: array<int32>, \n",
      "        AF_non_ukb_sas_XX: array<float64>, \n",
      "        AN_non_ukb_sas_XX: int32, \n",
      "        nhomalt_non_ukb_sas_XX: array<int32>, \n",
      "        AC_non_ukb_sas_XY: array<int32>, \n",
      "        AF_non_ukb_sas_XY: array<float64>, \n",
      "        AN_non_ukb_sas_XY: int32, \n",
      "        nhomalt_non_ukb_sas_XY: array<int32>, \n",
      "        AC_non_ukb_sas: array<int32>, \n",
      "        AF_non_ukb_sas: array<float64>, \n",
      "        AN_non_ukb_sas: int32, \n",
      "        nhomalt_non_ukb_sas: array<int32>, \n",
      "        AC_raw: array<int32>, \n",
      "        AF_raw: array<float64>, \n",
      "        AN_raw: int32, \n",
      "        nhomalt_raw: array<int32>, \n",
      "        AC_remaining_XX: array<int32>, \n",
      "        AF_remaining_XX: array<float64>, \n",
      "        AN_remaining_XX: int32, \n",
      "        nhomalt_remaining_XX: array<int32>, \n",
      "        AC_remaining_XY: array<int32>, \n",
      "        AF_remaining_XY: array<float64>, \n",
      "        AN_remaining_XY: int32, \n",
      "        nhomalt_remaining_XY: array<int32>, \n",
      "        AC_remaining: array<int32>, \n",
      "        AF_remaining: array<float64>, \n",
      "        AN_remaining: int32, \n",
      "        nhomalt_remaining: array<int32>, \n",
      "        AC_sas_XX: array<int32>, \n",
      "        AF_sas_XX: array<float64>, \n",
      "        AN_sas_XX: int32, \n",
      "        nhomalt_sas_XX: array<int32>, \n",
      "        AC_sas_XY: array<int32>, \n",
      "        AF_sas_XY: array<float64>, \n",
      "        AN_sas_XY: int32, \n",
      "        nhomalt_sas_XY: array<int32>, \n",
      "        AC_sas: array<int32>, \n",
      "        AF_sas: array<float64>, \n",
      "        AN_sas: int32, \n",
      "        nhomalt_sas: array<int32>, \n",
      "        AC_grpmax: array<int32>, \n",
      "        AF_grpmax: array<float64>, \n",
      "        AN_grpmax: array<int32>, \n",
      "        nhomalt_grpmax: array<int32>, \n",
      "        grpmax_non_ukb: array<str>, \n",
      "        AC_grpmax_non_ukb: array<int32>, \n",
      "        AF_grpmax_non_ukb: array<float64>, \n",
      "        AN_grpmax_non_ukb: array<int32>, \n",
      "        nhomalt_grpmax_non_ukb: array<int32>, \n",
      "        faf95_XX: array<float64>, \n",
      "        faf95_XY: array<float64>, \n",
      "        faf95: array<float64>, \n",
      "        faf95_afr_XX: array<float64>, \n",
      "        faf95_afr_XY: array<float64>, \n",
      "        faf95_afr: array<float64>, \n",
      "        faf95_amr_XX: array<float64>, \n",
      "        faf95_amr_XY: array<float64>, \n",
      "        faf95_amr: array<float64>, \n",
      "        faf95_eas_XX: array<float64>, \n",
      "        faf95_eas_XY: array<float64>, \n",
      "        faf95_eas: array<float64>, \n",
      "        faf95_mid_XX: array<float64>, \n",
      "        faf95_mid_XY: array<float64>, \n",
      "        faf95_mid: array<float64>, \n",
      "        faf95_nfe_XX: array<float64>, \n",
      "        faf95_nfe_XY: array<float64>, \n",
      "        faf95_nfe: array<float64>, \n",
      "        faf95_non_ukb_XX: array<float64>, \n",
      "        faf95_non_ukb_XY: array<float64>, \n",
      "        faf95_non_ukb: array<float64>, \n",
      "        faf95_non_ukb_afr_XX: array<float64>, \n",
      "        faf95_non_ukb_afr_XY: array<float64>, \n",
      "        faf95_non_ukb_afr: array<float64>, \n",
      "        faf95_non_ukb_amr_XX: array<float64>, \n",
      "        faf95_non_ukb_amr_XY: array<float64>, \n",
      "        faf95_non_ukb_amr: array<float64>, \n",
      "        faf95_non_ukb_eas_XX: array<float64>, \n",
      "        faf95_non_ukb_eas_XY: array<float64>, \n",
      "        faf95_non_ukb_eas: array<float64>, \n",
      "        faf95_non_ukb_mid_XX: array<float64>, \n",
      "        faf95_non_ukb_mid_XY: array<float64>, \n",
      "        faf95_non_ukb_mid: array<float64>, \n",
      "        faf95_non_ukb_nfe_XX: array<float64>, \n",
      "        faf95_non_ukb_nfe_XY: array<float64>, \n",
      "        faf95_non_ukb_nfe: array<float64>, \n",
      "        faf95_non_ukb_sas_XX: array<float64>, \n",
      "        faf95_non_ukb_sas_XY: array<float64>, \n",
      "        faf95_non_ukb_sas: array<float64>, \n",
      "        faf95_sas_XX: array<float64>, \n",
      "        faf95_sas_XY: array<float64>, \n",
      "        faf95_sas: array<float64>, \n",
      "        faf99_XX: array<float64>, \n",
      "        faf99_XY: array<float64>, \n",
      "        faf99: array<float64>, \n",
      "        faf99_afr_XX: array<float64>, \n",
      "        faf99_afr_XY: array<float64>, \n",
      "        faf99_afr: array<float64>, \n",
      "        faf99_amr_XX: array<float64>, \n",
      "        faf99_amr_XY: array<float64>, \n",
      "        faf99_amr: array<float64>, \n",
      "        faf99_eas_XX: array<float64>, \n",
      "        faf99_eas_XY: array<float64>, \n",
      "        faf99_eas: array<float64>, \n",
      "        faf99_mid_XX: array<float64>, \n",
      "        faf99_mid_XY: array<float64>, \n",
      "        faf99_mid: array<float64>, \n",
      "        faf99_nfe_XX: array<float64>, \n",
      "        faf99_nfe_XY: array<float64>, \n",
      "        faf99_nfe: array<float64>, \n",
      "        faf99_non_ukb_XX: array<float64>, \n",
      "        faf99_non_ukb_XY: array<float64>, \n",
      "        faf99_non_ukb: array<float64>, \n",
      "        faf99_non_ukb_afr_XX: array<float64>, \n",
      "        faf99_non_ukb_afr_XY: array<float64>, \n",
      "        faf99_non_ukb_afr: array<float64>, \n",
      "        faf99_non_ukb_amr_XX: array<float64>, \n",
      "        faf99_non_ukb_amr_XY: array<float64>, \n",
      "        faf99_non_ukb_amr: array<float64>, \n",
      "        faf99_non_ukb_eas_XX: array<float64>, \n",
      "        faf99_non_ukb_eas_XY: array<float64>, \n",
      "        faf99_non_ukb_eas: array<float64>, \n",
      "        faf99_non_ukb_mid_XX: array<float64>, \n",
      "        faf99_non_ukb_mid_XY: array<float64>, \n",
      "        faf99_non_ukb_mid: array<float64>, \n",
      "        faf99_non_ukb_nfe_XX: array<float64>, \n",
      "        faf99_non_ukb_nfe_XY: array<float64>, \n",
      "        faf99_non_ukb_nfe: array<float64>, \n",
      "        faf99_non_ukb_sas_XX: array<float64>, \n",
      "        faf99_non_ukb_sas_XY: array<float64>, \n",
      "        faf99_non_ukb_sas: array<float64>, \n",
      "        faf99_sas_XX: array<float64>, \n",
      "        faf99_sas_XY: array<float64>, \n",
      "        faf99_sas: array<float64>, \n",
      "        fafmax_faf99_max: array<float64>, \n",
      "        fafmax_faf99_max_gen_anc: array<str>, \n",
      "        fafmax_faf95_max_non_ukb: array<float64>, \n",
      "        fafmax_faf95_max_gen_anc_non_ukb: array<str>, \n",
      "        fafmax_faf99_max_non_ukb: array<float64>, \n",
      "        fafmax_faf99_max_gen_anc_non_ukb: array<str>, \n",
      "        age_hist_het_bin_freq: array<str>, \n",
      "        age_hist_het_n_smaller: array<int32>, \n",
      "        age_hist_het_n_larger: array<int32>, \n",
      "        age_hist_hom_bin_freq: array<str>, \n",
      "        age_hist_hom_n_smaller: array<int32>, \n",
      "        age_hist_hom_n_larger: array<int32>, \n",
      "        FS: float64, \n",
      "        MQ: float64, \n",
      "        MQRankSum: float64, \n",
      "        QUALapprox: int32, \n",
      "        QD: float64, \n",
      "        ReadPosRankSum: float64, \n",
      "        SOR: float64, \n",
      "        VarDP: int32, \n",
      "        monoallelic: bool, \n",
      "        only_het: bool, \n",
      "        transmitted_singleton: bool, \n",
      "        sibling_singleton: bool, \n",
      "        AS_FS: array<float64>, \n",
      "        AS_MQ: array<float64>, \n",
      "        AS_MQRankSum: array<float64>, \n",
      "        AS_pab_max: array<float64>, \n",
      "        AS_QUALapprox: array<int32>, \n",
      "        AS_QD: array<float64>, \n",
      "        AS_ReadPosRankSum: array<float64>, \n",
      "        AS_SB_TABLE: array<str>, \n",
      "        AS_SOR: array<float64>, \n",
      "        AS_VarDP: array<int32>, \n",
      "        inbreeding_coeff: array<float64>, \n",
      "        AS_culprit: array<str>, \n",
      "        AS_VQSLOD: array<float64>, \n",
      "        negative_train_site: bool, \n",
      "        positive_train_site: bool, \n",
      "        allele_type: str, \n",
      "        n_alt_alleles: int32, \n",
      "        variant_type: str, \n",
      "        was_mixed: bool, \n",
      "        lcr: bool, \n",
      "        non_par: bool, \n",
      "        segdup: bool, \n",
      "        fail_interval_qc: bool, \n",
      "        outside_ukb_capture_region: bool, \n",
      "        outside_broad_capture_region: bool, \n",
      "        gq_hist_alt_bin_freq: array<str>, \n",
      "        gq_hist_all_bin_freq: array<str>, \n",
      "        dp_hist_alt_bin_freq: array<str>, \n",
      "        dp_hist_alt_n_larger: array<int32>, \n",
      "        dp_hist_all_bin_freq: array<str>, \n",
      "        dp_hist_all_n_larger: array<int32>, \n",
      "        ab_hist_alt_bin_freq: array<str>, \n",
      "        cadd_raw_score: float64, \n",
      "        cadd_phred: float64, \n",
      "        revel_max: float64, \n",
      "        spliceai_ds_max: float64, \n",
      "        pangolin_largest_ds: float64, \n",
      "        phylop: float64, \n",
      "        sift_max: float64, \n",
      "        polyphen_max: float64, \n",
      "        VRS_Allele_IDs: array<str>, \n",
      "        VRS_Starts: array<int32>, \n",
      "        VRS_Ends: array<int32>, \n",
      "        VRS_States: array<str>, \n",
      "        vep: array<str>\n",
      "    }\n",
      "----------------------------------------\n",
      "Entry fields:\n",
      "    None\n",
      "----------------------------------------\n",
      "Column key: ['s']\n",
      "Row key: ['locus', 'alleles']\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Here need to be changed in the real dataset. The current dataset is just a mockup.\n",
    "# For example, adding annotations from gnomAD, filtering samples, etc.\n",
    "# # Calculate basic allele statistics (similar to gnomAD)\n",
    "# mt = hl.variant_qc(mt)\n",
    "print(\"Matrix table schema:\")\n",
    "mt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "79e0cfa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract gnomAD statistics from info field\n",
    "variants = mt.rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ac1d094c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper function to extract info field\n",
    "def first_or_default(expr, default):\n",
    "    return hl.if_else(hl.len(expr) > 0, expr[0], default)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d06a5fae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select relevant fields from the info struct\n",
    "variants = variants.select(\n",
    "    # Locus information\n",
    "    chrom=variants.locus.contig,\n",
    "    pos=variants.locus.position,\n",
    "    ref=variants.alleles[0],\n",
    "    alt=hl.delimit(variants.alleles[1:], ','),  # Join alt alleles if multiple\n",
    "    rsid=variants.rsid,\n",
    "\n",
    "    # Basic allele statistics (from gnomAD info field)\n",
    "    AC=first_or_default(variants.info.AC, 0),\n",
    "    AN=variants.info.AN,\n",
    "    AF=first_or_default(variants.info.AF, 0.0),\n",
    "\n",
    "    # Genetic ancestry group with max AF\n",
    "    grpmax=first_or_default(variants.info.grpmax, ''),\n",
    "\n",
    "    # Filtering Allele Frequency (95% confidence)\n",
    "    faf95_max=first_or_default(variants.info.fafmax_faf95_max, 0.0),\n",
    "    faf95_max_gen_anc=first_or_default(variants.info.fafmax_faf95_max_gen_anc, ''),\n",
    "\n",
    "    # Number of homozygotes\n",
    "    nhomalt=first_or_default(variants.info.nhomalt, 0),\n",
    "\n",
    "    # Sex-specific counts\n",
    "    AC_XX=first_or_default(variants.info.AC_XX, 0),\n",
    "    AN_XX=variants.info.AN_XX,\n",
    "    AF_XX=first_or_default(variants.info.AF_XX, 0.0),\n",
    "    nhomalt_XX=first_or_default(variants.info.nhomalt_XX, 0),\n",
    "\n",
    "    AC_XY=first_or_default(variants.info.AC_XY, 0),\n",
    "    AN_XY=variants.info.AN_XY,\n",
    "    AF_XY=first_or_default(variants.info.AF_XY, 0.0),\n",
    "    nhomalt_XY=first_or_default(variants.info.nhomalt_XY, 0),\n",
    "\n",
    "    # Population-specific (examples)\n",
    "    AC_afr=first_or_default(variants.info.AC_afr, 0),\n",
    "    AF_afr=first_or_default(variants.info.AF_afr, 0.0),\n",
    "    AN_afr=variants.info.AN_afr,\n",
    "\n",
    "    AC_amr=first_or_default(variants.info.AC_amr, 0),\n",
    "    AF_amr=first_or_default(variants.info.AF_amr, 0.0),\n",
    "    AN_amr=variants.info.AN_amr,\n",
    "\n",
    "    AC_eas=first_or_default(variants.info.AC_eas, 0),\n",
    "    AF_eas=first_or_default(variants.info.AF_eas, 0.0),\n",
    "    AN_eas=variants.info.AN_eas,\n",
    "\n",
    "    AC_nfe=first_or_default(variants.info.AC_nfe, 0),\n",
    "    AF_nfe=first_or_default(variants.info.AF_nfe, 0.0),\n",
    "    AN_nfe=variants.info.AN_nfe,\n",
    "\n",
    "    AC_sas=first_or_default(variants.info.AC_sas, 0),\n",
    "    AF_sas=first_or_default(variants.info.AF_sas, 0.0),\n",
    "    AN_sas=variants.info.AN_sas,\n",
    "\n",
    "    # Prediction scores\n",
    "    cadd_phred=variants.info.cadd_phred,\n",
    "    revel_max=variants.info.revel_max,\n",
    "    spliceai_ds_max=variants.info.spliceai_ds_max,\n",
    "\n",
    "    # Quality metrics\n",
    "    qual=variants.qual,\n",
    "    filters=hl.delimit(variants.filters, ','),\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8c5d81d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "=== Variant Statistics ===\n",
      "Total variants: 2,000\n",
      "\n",
      "First few variants:\n",
      "          locus       alleles  chrom     pos      ref  alt          rsid  \\\n",
      "0   chr18:48494        [T, A]  chr18   48494        T    A  rs1905848822   \n",
      "1   chr18:87087       [TG, T]  chr18   87087       TG    T          <NA>   \n",
      "2  chr18:109438        [T, C]  chr18  109438        T    C  rs1462125550   \n",
      "3  chr18:109567      [C, CCT]  chr18  109567        C  CCT          <NA>   \n",
      "4  chr18:112506  [CCTCCAG, C]  chr18  112506  CCTCCAG    C          <NA>   \n",
      "5  chr18:223630        [C, A]  chr18  223630        C    A          <NA>   \n",
      "6  chr18:224952        [C, A]  chr18  224952        C    A          <NA>   \n",
      "7  chr18:225235        [C, T]  chr18  225235        C    T  rs1911240230   \n",
      "8  chr18:225321        [T, C]  chr18  225321        T    C   rs755941042   \n",
      "9  chr18:260110        [A, G]  chr18  260110        A    G   rs540995383   \n",
      "\n",
      "     AC    AN        AF  ...    AF_nfe   AN_nfe AC_sas  AF_sas  AN_sas  \\\n",
      "0  <NA>  <NA>  0.000001  ...       0.0   627824      0     0.0   71926   \n",
      "1  <NA>  <NA>  0.000002  ...  0.000003   734590      0     0.0   71254   \n",
      "2  <NA>  <NA>       0.0  ...       0.0     1792      0     0.0      72   \n",
      "3  <NA>  <NA>       0.0  ...       0.0      758      0     0.0      10   \n",
      "4  <NA>  <NA>       0.0  ...       0.0   329612      0     0.0   65876   \n",
      "5  <NA>  <NA>       0.0  ...       0.0   370296      0     0.0   38710   \n",
      "6  <NA>  <NA>       0.0  ...       0.0  1090060      0     0.0   81342   \n",
      "7  <NA>  <NA>  0.000003  ...  0.000004  1057478      0     0.0   80016   \n",
      "8  <NA>  <NA>  0.000001  ...       0.0  1111064      0     0.0   86156   \n",
      "9  <NA>  <NA>  0.000007  ...       0.0   529738      0     0.0   28292   \n",
      "\n",
      "   cadd_phred  revel_max  spliceai_ds_max  qual      filters  \n",
      "0        5.98       <NA>             <NA>  <NA>      AS_VQSR  \n",
      "1       0.448       <NA>             <NA>  <NA>      AS_VQSR  \n",
      "2        10.8       <NA>             <NA>  <NA>  AC0,AS_VQSR  \n",
      "3       6.385       <NA>             <NA>  <NA>  AC0,AS_VQSR  \n",
      "4       5.847       <NA>             <NA>  <NA>  AC0,AS_VQSR  \n",
      "5       8.928       <NA>              0.0  <NA>          AC0  \n",
      "6        43.0       <NA>             0.05  <NA>          AC0  \n",
      "7        20.8       <NA>             0.09  <NA>               \n",
      "8        20.4       <NA>              0.0  <NA>               \n",
      "9        0.77       <NA>              0.0  <NA>               \n",
      "\n",
      "[10 rows x 42 columns]\n",
      "\n",
      "Exported to: ../data/variant_stats.tsv\n",
      "\n",
      "=== Summary Statistics ===\n",
      "Mean Allele Count: <NA>\n",
      "Mean Allele Frequency: 0.001223\n",
      "Mean Allele Number: <NA>\n",
      "Total Homozygotes: 784\n",
      "\n",
      "Chromosomes present: ['chr18', 'chr21']\n",
      "Variants with rsID: 593\n"
     ]
    }
   ],
   "source": [
    "# Convert to pandas\n",
    "df = variants.to_pandas()\n",
    "\n",
    "print(\"\\n=== Variant Statistics ===\")\n",
    "print(f\"Total variants: {len(df):,}\")\n",
    "print(\"\\nFirst few variants:\")\n",
    "print(df.head(10))\n",
    "\n",
    "# Export to TSV for Streamlit\n",
    "output_tsv = '../data/variant_stats.tsv'\n",
    "variants.export(output_tsv)\n",
    "print(f\"\\nExported to: {output_tsv}\")\n",
    "\n",
    "# Summary statistics\n",
    "print(\"\\n=== Summary Statistics ===\")\n",
    "print(f\"Mean Allele Count: {df['AC'].mean():.2f}\")\n",
    "print(f\"Mean Allele Frequency: {df['AF'].mean():.6f}\")\n",
    "print(f\"Mean Allele Number: {df['AN'].mean():.2f}\")\n",
    "print(f\"Total Homozygotes: {df['nhomalt'].sum():,}\")\n",
    "print(f\"\\nChromosomes present: {sorted(df['chrom'].unique())}\")\n",
    "print(f\"Variants with rsID: {df['rsid'].notna().sum():,}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gnomad-toolbox",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
